asterius/src/Asterius/Builtins/Primitive.hs

```diff
@@ -0,0 +1,183 @@
+import qualified Data.ByteString.Short as SBS
+
+-- GEORGE: Maybe use this function in all Asterius.Builtin.* modules to save us
+-- from duplication and copy-paste errors?
+
+-- | Create a function import with internal name of the form
+-- @__asterius_<EXTERNAL-MODULE-NAME>_<EXTERNAL-BASE-NAME>@.
+mkImport ::
+  -- | External module name
+  SBS.ShortByteString ->
+  -- | External base name
+  SBS.ShortByteString ->
+  -- | Function type.
+  FunctionType ->
+  FunctionImport
+mkImport ext_mod_name ext_base_name fn_type =
+  FunctionImport
+    { internalName = "__asterius_" <> ext_mod_name <> "_" <> ext_base_name,
+      externalModuleName = ext_mod_name,
+      externalBaseName = ext_base_name,
+      functionType = fn_type
+    }
+{-# INLINE mkImport #-}
+
+-- -------------------------------------------------------------------------
+
+-- GEORGE: Note that Asterius uses 64 bits everywhere, but WebAssembly at the
+-- moment does not support I64. Yet, it DOES support F64. Hence, we convert
+-- everything to F64 in all the signatures. It would be nice to remove this
+-- hack eventually, but that I guess is not in our hands at the moment.
+
+-- | Wasm import of the JavaScript implementation of @memcpy@, @memmove@,
+-- @memset@, and @memcmp@ (see implementations in rts/rts.memory.mjs). Notice
+-- that for some of them we ignore their result type; the hsprimitive_*
+-- functions do not use them.
+primitiveImports :: [FunctionImport]
+primitiveImports =
+  [ mkImport "Memory" "memcpy" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memmove" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memset" $
+      FunctionType {paramTypes = [F64, F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memsetFloat" $
+      FunctionType {paramTypes = [F64, F64, F64, F64], returnTypes = []},
+    mkImport "Memory" "memcmp" $
+      FunctionType {paramTypes = [F64, F64, F64], returnTypes = [F64]}
+  ]
+
+-- -------------------------------------------------------------------------
+
+primitiveCBits :: AsteriusModule
+primitiveCBits =
+  primitiveMemcpy
+    <> primitiveMemmove
+    <> primitiveMemcmp
+    <> primitiveMemsetWord8
+    <> primitiveMemsetWord16
+    <> primitiveMemsetWord32
+    <> primitiveMemsetWord64
+    <> primitiveMemsetWord
+    <> primitiveMemsetPtr
+    <> primitiveMemsetFloat
+    <> primitiveMemsetDouble
+    <> primitiveMemsetChar
+
+-- -------------------------------------------------------------------------
+
+-- | @void hsprimitive_memcpy(void *dst, ptrdiff_t doff, void *src, ptrdiff_t soff, size_t len)@
+primitiveMemcpy :: AsteriusModule
+primitiveMemcpy = runEDSL "hsprimitive_memcpy" $ do
+  setReturnTypes []
+  [dst, doff, src, soff, len] <- params [I64, I64, I64, I64, I64]
+  let arg1 = dst `addInt64` doff
+      arg2 = src `addInt64` soff
+  callImport "__asterius_Memory_memcpy" $
+    map convertSInt64ToFloat64 [arg1, arg2, len]
```

Comment:
In such cases, we're sure that both the addresses and the length values are non-negative, so it would be better to use `convertUInt64ToFloat64` instead.

Reply:
Ah right :+1: I changed all calls accordingly (only the value passed to memset now uses `convertSInt64ToFloat64`).